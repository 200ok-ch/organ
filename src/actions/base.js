import { displayFile, stopDisplayingFile } from './org';
import { sampleFileContents, whatsNewFileContents } from '../lib/static_file_contents';

import { Dropbox } from 'dropbox';

export const setLoadingMessage = loadingMessage => ({
  type: 'SET_LOADING_MESSAGE',
  loadingMessage,
});

export const hideLoadingMessage = () => ({
  type: 'HIDE_LOADING_MESSAGE',
});

export const setLastViewedFile = (lastViewedPath, lastViewedContents) => ({
  type: 'SET_LAST_VIEWED_FILE', lastViewedPath, lastViewedContents,
});

export const loadStaticFile = staticFile => {
  return (dispatch, getState) => {
    dispatch(setLastViewedFile(getState().org.present.get('path'),
                               getState().org.present.get('contents')));

    const fileContents = {
      'whats_new': whatsNewFileContents,
      'sample': sampleFileContents,
    }[staticFile];

    dispatch(displayFile(null, fileContents));
  };
};

export const unloadStaticFile = () => {
  return (dispatch, getState) => {
    dispatch(stopDisplayingFile());

    if (!!getState().base.get('lastViewedPath')) {
      dispatch(displayFile(getState().base.get('lastViewedPath'),
                           getState().base.get('lastViewedContents')));
    }
  };
};

export const setFontSize = newFontSize => ({
  type: 'SET_FONT_SIZE', newFontSize,
});

export const setBulletStyle = newBulletStyle => ({
  type: 'SET_BULLET_STYLE', newBulletStyle,
});

export const setShouldTapTodoToAdvance = newShouldTapTodoToAdvance => ({
  type: 'SET_SHOULD_TAP_TODO_TO_ADVANCE', newShouldTapTodoToAdvance,
});

export const setShouldStoreSettingsInDropbox = newShouldStoreSettingsInDropbox => {
  return (dispatch, getState) => {
    dispatch({ type: 'SET_SHOULD_STORE_SETTINGS_IN_DROPBOX', newShouldStoreSettingsInDropbox });

    if (!newShouldStoreSettingsInDropbox) {
      const dropbox = new Dropbox({ accessToken: getState().dropbox.get('accessToken') });
      dropbox.filesDelete({
        path: '/.org-web-config.json',
      }).catch(error => {
        if (!error.error.error['.tag'] === 'path_lookup') {
          alert(`There was an error trying to delete the .org-web-config.json file: ${error}`);
        }
      });

      window.previousSettingsFileContents = null;
    }
  };
};

export const setHasUnseenWhatsNew = newHasUnseenWhatsNew => ({
  type: 'SET_HAS_UNSEEN_WHATS_NEW', newHasUnseenWhatsNew,
});

export const setLastSeenWhatsNewHeader = newLastSeenWhatsNewHeader => ({
  type: 'SET_LAST_SEEN_WHATS_NEW_HEADER', newLastSeenWhatsNewHeader,
});

export const setCustomKeybinding = (keybindingName, keybinding) => ({
  type: 'SET_CUSTOM_KEYBINDING', keybindingName, keybinding,
});

export const restoreSettings = newSettings => ({
  type: 'RESTORE_SETTINGS', newSettings,
});

export const setActiveModalPage = activeModalPage => ({
  type: 'SET_ACTIVE_MODAL_PAGE', activeModalPage,
});
