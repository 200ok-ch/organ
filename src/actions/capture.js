export const addNewEmptyCaptureTemplate = () => ({
  type: 'ADD_NEW_EMPTY_CAPTURE_TEMPLATE',
});
