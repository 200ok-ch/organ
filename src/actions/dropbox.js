import { Dropbox } from 'dropbox';

import { fromJS } from 'immutable';

import { setLoadingMessage, hideLoadingMessage } from './base';

export const authenticate = accessToken => ({
  type: 'AUTHENTICATE',
  accessToken,
});

export const signOut = () => ({
  type: 'SIGN_OUT',
});

export const setCurrentFileBrowserDirectoryListing = (directoryPath, directoryListing) => {
  return {
    type: 'SET_CURRENT_FILE_BROWSER_DIRECTORY_LISTING',
    directoryPath,
    directoryListing
  };
};

export const getDirectoryListing = path => {
  return (dispatch, getState) => {
    const dropbox = new Dropbox({ accessToken: getState().dropbox.get('accessToken') });

    dispatch(setLoadingMessage('Getting listing...'));
    dropbox.filesListFolder({ path }).then(response => {
      const directoryListing = fromJS(response.entries.map(entry => ({
        id: entry.id,
        name: entry.name,
        isDirectory: entry['.tag'] === 'folder',
        path: entry.path_display,
      })));

      dispatch(setCurrentFileBrowserDirectoryListing(path, directoryListing));
      dispatch(hideLoadingMessage());
    }).catch(error => {
      console.error('There was an error retrieving files!');
      console.error(error);
    });
  };
};
