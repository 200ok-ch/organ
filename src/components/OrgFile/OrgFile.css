.org-file__parsing-error-message {
    text-align: center;
    font-size: 16px;

    margin: 10px;
}

.org-file__btn {
    margin: 10px;
    text-align: center;
}

.dirty-indicator {
    padding: 3px;
    background-color: gray;
    opacity: 0.5;
    color: white;
    position: fixed;
    bottom: 100px;
    right: 10px;
    font-weight: bold;
}

.error-message-container {
    padding: 10px;
    font-size: 20px;
}
