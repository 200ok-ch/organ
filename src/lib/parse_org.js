import generateId from './id_generator';

import { fromJS, List } from 'immutable';
import _ from 'lodash';

export const parseLinks = (rawText, { shouldAppendNewline = false } = {}) => {
  const linkRegex = /(\[\[([^\]]*)\]\]|\[\[([^\]]*)\]\[([^\]]*)\]\])/g;
  const matches = [];
  let match = linkRegex.exec(rawText);
  while (match) {
    if (match[2]) {
      matches.push({
        rawText: match[0],
        uri: match[2],
        index: match.index,
      });
    } else {
      matches.push({
        rawText: match[0],
        uri: match[3],
        title: match[4],
        index: match.index,
      });
    }
    match = linkRegex.exec(rawText);
  }

  const lineParts = [];
  let startIndex = 0;
  matches.forEach(match => {
    let index = match.index;

    if (index !== startIndex) {
      const text = rawText.substring(startIndex, index);
      lineParts.push({
        type: 'text',
        contents: text,
      });
    }

    const linkPart = {
      type: 'link',
      contents: {
        uri: match.uri,
      },
    };
    if (match.title) {
      linkPart.contents.title = match.title;
    }
    lineParts.push(linkPart);

    startIndex = match.index + match.rawText.length;
  });

  if (startIndex !== rawText.length || shouldAppendNewline) {
    const trailingText = rawText.substring(startIndex, rawText.length) + (shouldAppendNewline ? '\n' : '');
    lineParts.push({
      type: 'text',
      contents: trailingText,
    });
  }

  return lineParts;
};

const parseTable = tableLines => {
  const table = {
    id: generateId(),
    type: 'table',
    contents: [
      []
    ],
    columnProperties: [],
  };

  tableLines.map(line => line.trim()).forEach(line => {
    if (line.startsWith('|-')) {
      table.contents.push([]);
    } else {
      const lastRow = _.last(table.contents);
      const lineCells = line.substr(1, line.length - 2).split('|');

      if (lastRow.length === 0) {
        lineCells.forEach(cell => lastRow.push(cell));
      } else {
        lineCells.forEach((cellContents, cellIndex) => {
          lastRow[cellIndex] += `\n${cellContents}`;
        });
      }
    }
  });

  // Parse the contents of each cell.
  table.contents = table.contents.map(row => ({
    id: generateId(),
    contents: row.map(rawContents => ({
      id: generateId(),
      contents: parseLinks(rawContents),
      rawContents,
    }))
  }));

  // We sometimes end up with an extra, empty row - remove it if so.
  if (_.last(table.contents).contents.length === 0) {
    table.contents = table.contents.slice(0, table.contents.length - 1);
  }

  // Make sure each row has the same number of columns.
  const maxNumColumns = Math.max(...table.contents.map(row => row.contents.length));
  table.contents.forEach(row => {
    if (row.contents.length < maxNumColumns) {
      _.times(maxNumColumns - row.contents.length, () => {
        row.contents.push({
          id: generateId(),
          contents: [],
          rawContents: '',
        });
      });
    }
  });

  return table;
};

export const parseRawText = (rawText, { excludeContentElements = false } = {}) => {
  const lines = rawText.split('\n');

  const LIST_HEADER_REGEX = /^\s*([-+*]|(\d+(\.|\)))) (.*)/;

  let currentListHeaderNestingLevel = null;
  const rawLineParts = _.flatten(lines.map((line, lineIndex) => {
    const numLeadingSpaces = line.match(/^( *)/)[0].length;

    if (currentListHeaderNestingLevel !== null && (numLeadingSpaces > currentListHeaderNestingLevel || !line.trim())) {
      return [{
        type: 'raw-list-content', line,
      }];
    } else {
      currentListHeaderNestingLevel = null;

      if (!!line.match(LIST_HEADER_REGEX) && !excludeContentElements) {
        currentListHeaderNestingLevel = numLeadingSpaces;

        return [{
          type: 'raw-list-header', line,
        }];
      } else if (line.trim().startsWith('|') && !excludeContentElements) {
        return [{
          type: 'raw-table', line,
        }];
      } else {
        return parseLinks(line, { shouldAppendNewline: lineIndex !== lines.length - 1 });
      }
    }
  }));

  const processedLineParts = [];
  for (let partIndex = 0; partIndex < rawLineParts.length; ++partIndex) {
    const linePart = rawLineParts[partIndex];
    if (linePart.type === 'raw-table') {
      const tableLines = _.takeWhile(rawLineParts.slice(partIndex), part => (
        part.type === 'raw-table'
      )).map(part => part.line);

      processedLineParts.push(parseTable(tableLines));

      partIndex += tableLines.length - 1;
    } else if (linePart.type === 'raw-list-header') {
      const numLeadingSpaces = linePart.line.match(/^( *)/)[0].length;
      const contentLines = _.takeWhile(rawLineParts.slice(partIndex + 1), part => (
        part.type === 'raw-list-content'
      )).map(part => part.line).map(line => (
        line.startsWith(' '.repeat(numLeadingSpaces + 2)) ? (
          line.substr(numLeadingSpaces + 2)
        ) : (
          line.substr(numLeadingSpaces + 1)
        )
      ));
      if (contentLines[contentLines.length - 1] === '') {
        contentLines[contentLines.length - 1] = ' ';
      }
      const contents = parseRawText(contentLines.join('\n')).toJS();

      partIndex += contentLines.length;

      // Remove the leading -, +, *, or number characters.
      const line = linePart.line.match(LIST_HEADER_REGEX)[4];
      const newListItem = {
        id: generateId(),
        titleLine: parseLinks(line),
        contents,
      };

      const isOrdered = !!linePart.line.match(/\s*\d+[.)]/);

      const lastIndex = processedLineParts.length - 1;
      if (lastIndex >= 0 && processedLineParts[lastIndex].type === 'list') {
        processedLineParts[lastIndex].items.push(newListItem);
      } else {
        processedLineParts.push({
          type: 'list',
          id: generateId(),
          items: [newListItem],
          bulletCharacter: linePart.line.trim()[0],
          isOrdered,
        });
      }
    } else {
      processedLineParts.push(linePart);
    }
  }

  return fromJS(processedLineParts);
};

const defaultKeywordSets = fromJS([{
  keywords: ['TODO', 'DONE'],
  default: true
}]);

export const parseTitleLine = (titleLine, todoKeywordSets) => {
  const allKeywords = todoKeywordSets.flatMap(todoKeywordSet => {
    return todoKeywordSet.get('keywords');
  });
  const todoKeyword = allKeywords.filter(keyword => titleLine.startsWith(keyword + ' ')).first();
  let rawTitle = titleLine;
  if (todoKeyword) {
    rawTitle = rawTitle.substr(todoKeyword.length + 1);
  }

  // Check for tags.
  let tags = [];
  if (rawTitle.trimRight().endsWith(':')) {
    const titleParts = rawTitle.trimRight().split(' ');
    const possibleTags = titleParts[titleParts.length - 1];
    if (/^:[^\s]+:$/.test(possibleTags)) {
      rawTitle = rawTitle.substr(0, rawTitle.length - possibleTags.length);
      tags = possibleTags.split(':').filter(tag => tag !== '');
    }
  }

  const title = parseRawText(rawTitle, { excludeContentElements: true });

  return fromJS({ title, rawTitle, todoKeyword, tags });
};

export const newHeaderWithTitle = (line, nestingLevel, todoKeywordSets) => {
  if (todoKeywordSets.size === 0) {
    todoKeywordSets = defaultKeywordSets;
  }

  const titleLine = parseTitleLine(line, todoKeywordSets);
  return fromJS({
    titleLine,
    rawDescription: '',
    description: [],
    opened: false,
    id: generateId(),

    nestingLevel
  });
};

export const newHeaderFromText = (rawText, todoKeywordSets) => {
  const titleLine = rawText.split('\n')[0].replace(/^\**\s*/, '');
  const description = rawText.split('\n').slice(1).join('\n');

  return newHeaderWithTitle(titleLine, 1, todoKeywordSets)
    .set('rawDescription', description)
    .set('description', parseRawText(description));
};

export const parseOrg = (fileContents) => {
  let headers = new List();
  const lines = fileContents.split('\n');

  let todoKeywordSets = new List();

  lines.forEach(line => {
    if (line.startsWith('*')) {
      let nestingLevel = line.indexOf(' ');
      if (nestingLevel === -1) {
        nestingLevel = line.length;
      }
      const title = line.substr(nestingLevel + 1);
      headers = headers.push(newHeaderWithTitle(title, nestingLevel, todoKeywordSets));
    } else {
      if (headers.size === 0) {
        if (line.startsWith('#+TODO: ') || line.startsWith('#+TYP_TODO: ')) {
          const keywordsString = line.substr(line.indexOf(':') + 2);
          const keywordStrings = keywordsString.split(/\s/).filter(keyword => {
            return keyword !== '|';
          });
          const keywords = keywordStrings.map(keywordString => {
            const todoRegex = /([^(]*)(\(.*\))?/g;
            const match = todoRegex.exec(keywordString);
            const keyword = match[1];

            return keyword;
          });
          todoKeywordSets = todoKeywordSets.push(fromJS({
            keywords,
            configLine: line,
            default: false
          }));
        }
      } else {
        headers = headers.updateIn([headers.size - 1, 'rawDescription'], rawDescription => (
          rawDescription.length === 0 ? line : rawDescription + '\n' + line
        ));
      }
    }
  });

  if (todoKeywordSets.size === 0) {
    todoKeywordSets = defaultKeywordSets;
  }

  headers = headers.map(header => {
    return header.set('description', parseRawText(header.get('rawDescription')));
  });

  return fromJS({
    headers, todoKeywordSets
  });
};
