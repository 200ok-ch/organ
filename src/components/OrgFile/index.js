import React, { PureComponent, Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { Redirect } from 'react-router-dom';

import './OrgFile.css';

import HeaderList from './components/HeaderList';
import ActionDrawer from './components/ActionDrawer';

import * as baseActions from '../../actions/base';
import * as dropboxActions from '../../actions/dropbox';

class OrgFile extends PureComponent {
  componentDidMount() {
    const { staticFile, path } = this.props;

    if (!!staticFile) {
      this.props.base.loadStaticFile(staticFile);
    } else if (!!path) {
      this.props.dropbox.downloadFile(path);
    }
  }

  componentWillUnmount() {
    const { staticFile } = this.props;

    if (!!staticFile) {
      this.props.base.unloadStaticFile();
    }
  }

  render() {
    const {
      headers,
      backButtonText,
      onBackClick,
      shouldDisableDirtyIndicator,
      shouldDisableSyncButtons,
      shouldDisableActionDrawer,
      isDirty,
      parsingErrorMessage,
      path,
      staticFile,
    } = this.props;

    if (!path && !staticFile) {
      return <Redirect to="/files" />;
    }

    if (!headers) {
      return <div></div>;
    }

    return (
      <div>
        {headers.size === 0 ? (
          <div className="org-file__parsing-error-message">
            <h3>Couldn't parse file</h3>

            {!!parsingErrorMessage ? (
              <Fragment>{parsingErrorMessage}</Fragment>
            ) : (
              <Fragment>
                If you think this is a bug, please
                {' '}<a href="https://github.com/DanielDe/org-web/issues/new" target="_blank" rel="noopener noreferrer">create an issue</a>
                {' '}and include the org file if possible!
              </Fragment>
            )}
          </div>
        ) : (
          <HeaderList />
        )}

        <div className="btn org-file__btn" onClick={onBackClick}>
          {backButtonText}
        </div>

        {isDirty && !shouldDisableDirtyIndicator && <div className="dirty-indicator">Unpushed changes</div>}

        {!shouldDisableActionDrawer && <ActionDrawer shouldDisableSyncButtons={shouldDisableSyncButtons} />}
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    headers: state.org.present.get('headers'),
    isDirty: state.org.present.get('isDirty'),
  };
};

const mapDispatchToProps = dispatch => {
  return {
    base: bindActionCreators(baseActions, dispatch),
    dropbox: bindActionCreators(dropboxActions, dispatch),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(OrgFile);
