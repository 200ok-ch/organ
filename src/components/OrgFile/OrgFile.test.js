import React from 'react';
import thunk from 'redux-thunk';

import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';

import OrgFile from './';

import { parseOrg, _parsePlanningItems } from '../../lib/parse_org';
import exportOrg from '../../lib/export_org';
import rootReducer from '../../reducers/';

import { displayFile } from '../../actions/org';

import { Map, fromJS } from 'immutable';

import readFixture from '../../../test_helpers/index';
import { render, fireEvent, cleanup } from '@testing-library/react';
/**
 * This is a convenience wrapper around paring an org file using
 * `parseOrg` and then export it using `exportOrg`.
 * @param {String} testOrgFile - contents of an org file
 */
function parseAndExportOrgFile(testOrgFile) {
  const parsedFile = parseOrg(testOrgFile);
  const headers = parsedFile.get('headers');
  const todoKeywordSets = parsedFile.get('todoKeywordSets');
  const exportedFile = exportOrg(headers, todoKeywordSets);
  return exportedFile;
}

afterEach(cleanup);

describe('Unit Tests for org file', () => {
  describe('Parsing', () => {
    test("Parsing and exporting shouldn't alter the original file", () => {
      const testOrgFile = readFixture('indented_list');
      const exportedFile = parseAndExportOrgFile(testOrgFile);

      // Should have the same amount of lines. Safeguard for the next
      // expectation.
      const exportedFileLines = exportedFile.split('\n');
      const testOrgFileLines = testOrgFile.split('\n');
      expect(exportedFileLines.length).toEqual(testOrgFileLines.length);

      exportedFileLines.forEach((line, index) => {
        expect(line).toEqual(testOrgFileLines[index]);
      });
    });

    // TODO: Write such a file
    test.skip('Parses and exports a file which contains all features of organice', () => {});

    describe('Boldness', () => {
      test('Parsing lines with bold text', () => {
        const testOrgFile = readFixture('bold_text');
        const exportedFile = parseAndExportOrgFile(testOrgFile);
        expect(exportedFile).toEqual(testOrgFile);
      });
    });

    describe('Newlines', () => {
      test('Newlines in between headers and items are preserved', () => {
        const testOrgFile = readFixture('newlines');
        const exportedFile = parseAndExportOrgFile(testOrgFile);
        expect(exportedFile).toEqual(testOrgFile);
      });
    });

    describe('Planning items', () => {
      describe('Formatting is the same as in Emacs', () => {
        describe('List formatting', () => {
          test('Parsing a basic list should not mangle the list', () => {
            const testDescription = '  - indented list\n     - Foo';
            const parsedFile = _parsePlanningItems(testDescription);
            expect(parsedFile.strippedDescription).toEqual(testDescription);
          });

          test('Parsing a list with planning items should not mangle the list', () => {
            const testDescription = '  - indented list\n     - Foo';
            const parsedFile = _parsePlanningItems(
              `SCHEDULED: <2019-07-30 Tue>\n${testDescription}`
            );
            expect(parsedFile.strippedDescription).toEqual(testDescription);
          });
        });

        describe('Planning items are formatted as is default Emacs', () => {
          test('For basic files', () => {
            const testOrgFile = readFixture('schedule');
            const exportedFile = parseAndExportOrgFile(testOrgFile);
            // The call to `trimRight` is a work-around, because organice
            // doesn't export files with a trailing newline at this
            // moment. This is best-practice for any text-file and Emacs
            // does it for org-files, too. However, this is to be fixed
            // at another time. And when it is, this expectation will
            // fail and the call to `trimRight` can be safely removed.
            expect(exportedFile).toEqual(testOrgFile.trimRight());
          });

          test('For files with multiple planning items', () => {
            const testOrgFile = readFixture('schedule_and_deadline');
            const exportedFile = parseAndExportOrgFile(testOrgFile);
            expect(exportedFile).toEqual(testOrgFile.trimRight());
          });
        });

        test('Properties are formatted as is default in Emacs', () => {
          const testOrgFile = readFixture('properties');
          const exportedFile = parseAndExportOrgFile(testOrgFile);
          expect(exportedFile).toEqual(testOrgFile);
        });

        test('Tags are formatted as is default in Emacs', () => {
          const testOrgFile = readFixture('tags');
          const exportedFile = parseAndExportOrgFile(testOrgFile);
          expect(exportedFile).toEqual(testOrgFile);
        });
      });
    });
  });
});

describe('Rendering an org file', () => {
  jest.mock('react-hotkeys', () => {
    const React = require('react');
    const Fragment = React.Fragment;

    return {
      HotKeys: ({ children }) => <Fragment>{children}</Fragment>,
    };
  });

  const testOrgFile = `
* Top level header
** A nested header
** TODO A todo item
* Another top level header
Some description content
* A header with tags                                              :tag1:tag2:
* A header with [[https://organice.200ok.ch][a link]]
`;

  let store;

  beforeEach(() => {
    let capture = new Map();
    capture = capture.set('captureTemplates', []);
    store = createStore(
      rootReducer,
      {
        org: {
          past: [],
          present: new Map(),
          future: [],
        },
        syncBackend: Map(),
        capture,
        base: new fromJS({
          customKeybindings: {},
          shouldTapTodoToAdvance: true,
        }),
      },
      applyMiddleware(thunk)
    );
    store.dispatch(displayFile('fixtureTestFile.org', testOrgFile));
  });

  test('<OrgFile /> renders an org file', () => {
    const { container, getAllByText } = render(
      <MemoryRouter keyLength={0}>
        <Provider store={store}>
          <OrgFile path="fixtureTestFile.org" />
        </Provider>
      </MemoryRouter>
    );

    expect(getAllByText(/\*/)).toHaveLength(4);
    expect(container).toMatchSnapshot();
  });

  test('Can select a header in an org file', () => {
    const { getByText, queryByText } = render(
      <MemoryRouter keyLength={0}>
        <Provider store={store}>
          <OrgFile path="fixtureTestFile.org" />
        </Provider>
      </MemoryRouter>
    );

    expect(queryByText('Scheduled')).toBeFalsy();
    expect(queryByText('Deadline')).toBeFalsy();

    fireEvent.click(getByText('Top level header'));

    expect(queryByText('Scheduled')).toBeTruthy();
    expect(queryByText('Deadline')).toBeTruthy();
  });

  // Org Mode has keywords as workflow states and can cycle through
  // them: https://orgmode.org/manual/Workflow-states.html
  // In organice, we can cycle through them by swiping or by clicking
  // (if enabled). This test checks for the latter.
  test('Can advance todo state for selected header in an org file', () => {
    const { queryByText } = render(
      <MemoryRouter keyLength={0}>
        <Provider store={store}>
          <OrgFile path="fixtureTestFile.org" />
        </Provider>
      </MemoryRouter>
    );

    // In the very beginning, the TODO is hidden, because the file
    // starts folded down to the top level
    expect(queryByText('TODO')).toBeFalsy();

    // Cycle the top header (which will make the next headers
    // [including the TODO] visible
    fireEvent.click(queryByText('Top level header'));

    // In the beginning, the TODO is not DONE
    expect(queryByText('TODO')).toBeTruthy();
    expect(queryByText('DONE')).toBeFalsy();

    // Toggle the TODO
    fireEvent.click(queryByText('TODO'));

    // Then the TODO is DONE
    expect(queryByText('TODO')).toBeFalsy();
    expect(queryByText('DONE')).toBeTruthy();
  });
});
