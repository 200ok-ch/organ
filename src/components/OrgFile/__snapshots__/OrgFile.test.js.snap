// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering an org file <OrgFile /> renders an org file 1`] = `
<div>
  <div
    tabindex="-1"
  >
    <div
      class="org-file-container"
      tabindex="-1"
    >
      <div
        class="header-list-container"
      >
        <div
          class="header"
          style="padding-left: 20px; margin-left: 0px;"
        >
          <div
            class="left-swipe-action-container"
            style="width: 0px; background-color: rgb(211, 211, 211);"
          >
            <i
              class="fas fa-check swipe-action-container__icon swipe-action-container__icon--left"
              style="display: none;"
            />
          </div>
          <div
            class="right-swipe-action-container"
            style="width: 0px; background-color: rgb(211, 211, 211);"
          >
            <i
              class="fas fa-times swipe-action-container__icon swipe-action-container__icon--right"
              style="display: none;"
            />
          </div>
          <div
            class="header__bullet"
            style="margin-left: -16px;"
          >
            *
          </div>
          <div
            class="title-line"
            style="width: 0px;"
          >
            
            <div>
              <span
                style="word-break: break-word;"
              >
                <span>
                  Top level header
                </span>
                ...
              </span>
            </div>
          </div>
          <div />
        </div>
        <div
          class="header"
          style="padding-left: 20px; margin-left: 0px;"
        >
          <div
            class="left-swipe-action-container"
            style="width: 0px; background-color: rgb(211, 211, 211);"
          >
            <i
              class="fas fa-check swipe-action-container__icon swipe-action-container__icon--left"
              style="display: none;"
            />
          </div>
          <div
            class="right-swipe-action-container"
            style="width: 0px; background-color: rgb(211, 211, 211);"
          >
            <i
              class="fas fa-times swipe-action-container__icon swipe-action-container__icon--right"
              style="display: none;"
            />
          </div>
          <div
            class="header__bullet"
            style="margin-left: -16px;"
          >
            *
          </div>
          <div
            class="title-line"
            style="width: 0px;"
          >
            
            <div>
              <span
                style="word-break: break-word;"
              >
                <span>
                  Another top level header
                </span>
                ...
              </span>
            </div>
          </div>
          <div />
        </div>
        <div
          class="header"
          style="padding-left: 20px; margin-left: 0px;"
        >
          <div
            class="left-swipe-action-container"
            style="width: 0px; background-color: rgb(211, 211, 211);"
          >
            <i
              class="fas fa-check swipe-action-container__icon swipe-action-container__icon--left"
              style="display: none;"
            />
          </div>
          <div
            class="right-swipe-action-container"
            style="width: 0px; background-color: rgb(211, 211, 211);"
          >
            <i
              class="fas fa-times swipe-action-container__icon swipe-action-container__icon--right"
              style="display: none;"
            />
          </div>
          <div
            class="header__bullet"
            style="margin-left: -16px;"
          >
            *
          </div>
          <div
            class="title-line"
            style="width: 0px;"
          >
            
            <div>
              <span
                style="word-break: break-word;"
              >
                <span>
                  A header with tags                                              
                </span>
                
              </span>
              <div>
                <div
                  class="header-tag"
                >
                  tag1
                </div>
                <div
                  class="header-tag"
                >
                  tag2
                </div>
              </div>
            </div>
          </div>
          <div />
        </div>
        <div
          class="header"
          style="padding-left: 20px; margin-left: 0px;"
        >
          <div
            class="left-swipe-action-container"
            style="width: 0px; background-color: rgb(211, 211, 211);"
          >
            <i
              class="fas fa-check swipe-action-container__icon swipe-action-container__icon--left"
              style="display: none;"
            />
          </div>
          <div
            class="right-swipe-action-container"
            style="width: 0px; background-color: rgb(211, 211, 211);"
          >
            <i
              class="fas fa-times swipe-action-container__icon swipe-action-container__icon--right"
              style="display: none;"
            />
          </div>
          <div
            class="header__bullet"
            style="margin-left: -16px;"
          >
            *
          </div>
          <div
            class="title-line"
            style="width: 0px;"
          >
            
            <div>
              <span
                style="word-break: break-word;"
              >
                <span>
                  A header with 
                  <a
                    href="https://organice.200ok.ch"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    a link
                  </a>
                </span>
                ...
              </span>
            </div>
          </div>
          <div />
        </div>
      </div>
      <div
        class="action-drawer-container nice-scroll"
      >
        <button
          class="btn btn--circle action-drawer__btn fas fa-lg fa-cloud action-drawer__btn--with-sub-icon"
          style="opacity: 1;"
          title="Sync changes"
        >
          <i
            class="fas fa-xs fa-sync-alt action-drawer__btn__sub-icon"
          />
        </button>
        <div
          class="action-drawer__arrow-buttons-container"
          style="left: 0px;"
        >
          <button
            class="btn btn--circle action-drawer__btn action-drawer__arrow-button fas fa-lg fa-arrow-up"
            style="opacity: 1; box-shadow: none; bottom: 0px;"
            title="Move header up"
          />
          <button
            class="btn btn--circle action-drawer__btn action-drawer__arrow-button fas fa-lg fa-arrow-down"
            style="opacity: 1; box-shadow: none; bottom: 0px;"
            title="Move header down"
          />
          <button
            class="btn btn--circle action-drawer__btn action-drawer__arrow-button fas fa-lg fa-arrow-left"
            style="opacity: 1; box-shadow: none; bottom: 0px; right: 0px;"
            title="Move header left"
          />
          <button
            class="btn btn--circle action-drawer__btn action-drawer__arrow-button fas fa-lg fa-arrow-right"
            style="opacity: 1; box-shadow: none; bottom: 0px; left: 0px;"
            title="Move header right"
          />
          <button
            class="btn btn--circle action-drawer__btn action-drawer__arrow-button fas fa-lg fa-chevron-left"
            style="opacity: 1; box-shadow: none; bottom: 0px; right: 0px;"
            title="Move entire subtree left"
          />
          <button
            class="btn btn--circle action-drawer__btn action-drawer__arrow-button fas fa-lg fa-chevron-right"
            style="opacity: 1; box-shadow: none; bottom: 0px; left: 0px;"
            title="Move entire subtree right"
          />
          <button
            class="btn btn--circle action-drawer__btn action-drawer__main-arrow-button fas fa-lg fa-arrows-alt"
            style="opacity: 1;"
            title="Show movement buttons"
          />
        </div>
        <button
          class="btn btn--circle action-drawer__btn fas fa-lg fa-calendar-alt"
          style="opacity: 1;"
          title="Show agenda"
        />
        <div
          class="action-drawer__capture-buttons-container"
        >
          <button
            class="btn btn--circle action-drawer__btn fas fa-lg fa-list-ul"
            style="opacity: 1; position: relative; z-index: 1;"
            title="Show capture templates"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
